Glide.get fails if PackageManager IPC fails

PackageManager has a failure mode where it throws a RuntimeException with "Package manager has died" when IPC calls fail.  We see this in the Dropbox app quite frequently.  In our internal calls to PackageManager we are careful to catch this RuntimeException and handle it in cases where we can.  

Unfortunately, adding Glide to our dependency list has started showing crashes with this stack:

``` ruby
Caused by: java.lang.RuntimeException: Package manager has died
    at android.app.ApplicationPackageManager.getApplicationInfo(ApplicationPackageManager.java:250)
    at com.bumptech.glide.module.ManifestParser.parse(panda.py:25)
    at com.bumptech.glide.Glide.get(panda.py:147)
    at com.dropbox.android.taskqueue.AnimatedGifDownloader.<init>(panda.py:139)
    at com.dropbox.android.DropboxApplication$AppState.initializeMainProcess(panda.py:899)
    at com.dropbox.android.DropboxApplication$AppState.<init>(panda.py:681)
    at com.dropbox.android.DropboxApplication.onCreate(panda.py:207)
```

For some reason not quite clear to me, we don't see this on anything above 4.4.4.

The request here would be to have a reasonable fallback when PackageManager is unavailable, or a way for apps to opt out of manifest scanning to pick up custom modules, since we don't use any.

--- a/library/src/androidTest/java/com/bumptech/glide/GlideTest.java
+++ b/library/src/androidTest/java/com/bumptech/glide/GlideTest.java
@@ -1,5 +1,6 @@
 package com.bumptech.glide;
 
+import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
 import static org.mockito.Matchers.any;
 import static org.mockito.Matchers.anyInt;
@@ -14,6 +15,7 @@ import static org.mockito.Mockito.when;
 
 import android.content.ContentResolver;
 import android.content.Context;
+import android.content.pm.ApplicationInfo;
 import android.content.res.AssetFileDescriptor;
 import android.graphics.Bitmap;
 import android.graphics.Color;
@@ -21,6 +23,7 @@ import android.graphics.drawable.ColorDrawable;
 import android.graphics.drawable.Drawable;
 import android.media.MediaMetadataRetriever;
 import android.net.Uri;
+import android.os.Bundle;
 import android.os.Handler;
 import android.os.ParcelFileDescriptor;
 import android.view.ViewGroup;
@@ -47,6 +50,7 @@ import com.bumptech.glide.load.resource.gifbitmap.GifBitmapWrapper;
 import com.bumptech.glide.load.resource.transcode.ResourceTranscoder;
 import com.bumptech.glide.manager.Lifecycle;
 import com.bumptech.glide.manager.RequestManagerTreeNode;
+import com.bumptech.glide.module.GlideModule;
 import com.bumptech.glide.request.Request;
 import com.bumptech.glide.request.RequestListener;
 import com.bumptech.glide.request.animation.GlideAnimation;
@@ -67,6 +71,7 @@ import org.robolectric.annotation.Config;
 import org.robolectric.annotation.Implementation;
 import org.robolectric.annotation.Implements;
 import org.robolectric.annotation.Resetter;
+import org.robolectric.res.builder.RobolectricPackageManager;
 import org.robolectric.shadows.ShadowBitmap;
 
 import java.io.ByteArrayInputStream;
@@ -713,6 +718,74 @@ public class GlideTest {
 
         secondRequest.load(fakeUri).into(secondTarget);
         verify(secondTarget).onResourceReady(eq(secondResult), any(GlideAnimation.class));
+    }
+
+    @Test
+    public void testSetModulesEnabledTrue() throws Exception {
+        // tearDown glide instance first so we have a clean slate and not the Glide.get() call in setUp
+        Glide.tearDown();
+
+        // Register a module in the manifest
+        setupMockModule();
+
+        // Reset the static counters
+        MockModule.applyOptionsCallCount = MockModule.registerComponentsCallCount = 0;
+
+        // This is the default, so it should be a no-op
+        Glide.setModulesEnabled(true);
+
+        Glide.get(getContext());
+
+        // With modules enabled, MockModule below should have been initialized,
+        // so the counters should both be 1
+        assertEquals(1, MockModule.registerComponentsCallCount);
+        assertEquals(1, MockModule.applyOptionsCallCount);
+    }
+
+    @Test
+    public void testSetModulesEnabledFalse() throws Exception {
+        // tearDown glide instance first so we have a clean slate and not the Glide.get() call in setUp
+        Glide.tearDown();
+
+        // Register a module in the manifest
+        setupMockModule();
+
+        // Reset the static counters
+        MockModule.applyOptionsCallCount = MockModule.registerComponentsCallCount = 0;
+
+        Glide.setModulesEnabled(false);
+
+        Glide.get(getContext());
+
+        // With modules disabled, MockModule below should not have been initialized,
+        // so the counters should still be 0
+        assertEquals(0, MockModule.registerComponentsCallCount);
+        assertEquals(0, MockModule.applyOptionsCallCount);
+    }
+
+    private void setupMockModule() throws Exception {
+        RobolectricPackageManager pm =
+                (RobolectricPackageManager) Robolectric.application.getPackageManager();
+        ApplicationInfo info =
+                pm.getApplicationInfo(Robolectric.application.getPackageName(), 0);
+        info.metaData = new Bundle();
+        info.metaData.putString(MockModule.class.getName(), "GlideModule");
+    }
+
+    public static class MockModule implements GlideModule {
+
+        static int applyOptionsCallCount;
+        static int registerComponentsCallCount;
+
+        @Override
+        public void applyOptions(Context context, GlideBuilder builder) {
+            applyOptionsCallCount++;
+        }
+
+        @Override
+        public void registerComponents(Context context, Glide glide) {
+            registerComponentsCallCount++;
+        }
     }
 
     @SuppressWarnings("unchecked")

--- a/library/src/main/java/com/bumptech/glide/Glide.java
+++ b/library/src/main/java/com/bumptech/glide/Glide.java
@@ -72,6 +72,7 @@ import com.bumptech.glide.util.Util;
 import java.io.File;
 import java.io.InputStream;
 import java.net.URL;
+import java.util.Collections;
 import java.util.List;
 
 /**
@@ -83,6 +84,7 @@ public class Glide {
 
     private static final String TAG = "Glide";
     private static volatile Glide glide;
+    private static boolean modulesEnabled = true;
 
     private final GenericLoaderFactory loaderFactory;
     private final Engine engine;
@@ -137,6 +139,21 @@ public class Glide {
     }
 
     /**
+     * Enable or disable the parsing of AndroidManifest.xml
+     * looking for {@link GlideModule} implementations.
+     * @throws IllegalArgumentException if the Glide singleton has already been created.
+     */
+    public static void setModulesEnabled(boolean enabled) {
+        synchronized (Glide.class) {
+            if (glide != null) {
+                throw new IllegalArgumentException("Glide singleton already exists.");
+            }
+
+            modulesEnabled = enabled;
+        }
+    }
+
+    /**
      * Get the singleton.
      *
      * @return the singleton
@@ -146,9 +163,8 @@ public class Glide {
             synchronized (Glide.class) {
                 if (glide == null) {
                     Context applicationContext = context.getApplicationContext();
-                    List<GlideModule> modules = new ManifestParser(applicationContext).parse();
-
                     GlideBuilder builder = new GlideBuilder(applicationContext);
+                    List<GlideModule> modules = parseGlideModules(applicationContext);
                     for (GlideModule module : modules) {
                         module.applyOptions(applicationContext, builder);
                     }
@@ -161,6 +177,18 @@ public class Glide {
         }
 
         return glide;
+    }
+
+    /**
+     * If modules are enabled, parses the application manifest and returns the configured modules.
+     * Otherwise, returns an empty list.
+     */
+    private static List<GlideModule> parseGlideModules(Context applicationContext) {
+        if (modulesEnabled) {
+            return new ManifestParser(applicationContext).parse();
+        } else {
+            return Collections.emptyList();
+        }
     }
 
     /**
@@ -198,6 +226,7 @@ public class Glide {
     // For testing.
     static void tearDown() {
         glide = null;
+        modulesEnabled = true;
     }
 
     Glide(Engine engine, MemoryCache memoryCache, BitmapPool bitmapPool, Context context, DecodeFormat decodeFormat) {

--- a/library/src/main/java/com/bumptech/glide/Glide.java
+++ b/library/src/main/java/com/bumptech/glide/Glide.java
@@ -139,8 +139,7 @@ public class Glide {
     }
 
     /**
-     * Enable or disable the parsing of AndroidManifest.xml
-     * looking for {@link GlideModule} implementations.
+     * Enable or disable the parsing of AndroidManifest.xml looking for {@link GlideModule} implementations.
      * @throws IllegalArgumentException if the Glide singleton has already been created.
      */
     public static void setModulesEnabled(boolean enabled) {

